File: python/gtk4/oop/action-bar.py
import gi, sys
gi.require_version("Gtk", "4.0")
from gi.repository import Gtk

# GtkActionBar: A full width container to add contextual actions

class App(Gtk.Application):
    APP_ID = "io.github.Miqueas.GTK-Examples.Python.Gtk4.OOP.ActionBar"
    APP_TITLE = "Gtk.ActionBar"

    def __init__(self):
        super().__init__(application_id = self.APP_ID)

    def do_activate(self):
        window = self.get_active_window()
        if window: window.present()

    def do_startup(self):
        # Ensures chain-up
        Gtk.Application.do_startup(self)
        
        window = Gtk.ApplicationWindow(application = self)
        box = Gtk.Box(orientation = Gtk.Orientation.VERTICAL)
        label = Gtk.Label(label = "App content here", vexpand = True, hexpand = True)
        actionBar = Gtk.ActionBar()

        window.set_child(box)
        window.set_title(self.APP_TITLE)
        window.set_default_size(400, 400)

        box.append(label)
        box.append(actionBar)

        actionBar.pack_start(Gtk.Label(label = "Something"))
        actionBar.pack_end(Gtk.Button(label = "A button"))

if __name__ == "__main__":
    sys.exit(App().run(sys.argv))

File: cpp/gtk4/oop/action-bar.cpp
#include <gtkmm.h>

static const Glib::ustring APP_ID = "io.github.Miqueas.GTK-Examples.Cpp.Gtk4.OOP.ActionBar";
static const Glib::ustring APP_TITLE = "Gtk::ActionBar";

class App : public Gtk::Application {
  public:
    static Glib::RefPtr<App> create() {
      return Glib::make_refptr_for_instance<App>(new App());
    }

  protected:
    App() : Gtk::Application(APP_ID) {}

    void on_activate() override {
      Gtk::Window* window = get_active_window();
      if (window != nullptr) window->present();
    }

    void on_startup() override {
      // Ensures chain-up
      Gtk::Application::on_startup();

      auto window = Gtk::make_managed<Gtk::ApplicationWindow>();
      auto action_bar = Gtk::make_managed<Gtk::ActionBar>();
      auto action_bar_label = Gtk::make_managed<Gtk::Label>("Something");
      auto action_bar_button = Gtk::make_managed<Gtk::Button>("A button");
      auto box = Gtk::make_managed<Gtk::Box>(Gtk::Orientation::VERTICAL, 0);
      auto box_label = Gtk::make_managed<Gtk::Label>("App content");

      window->set_child(*box);
      window->set_title(APP_TITLE);
      window->set_default_size(400, 400);

      action_bar->pack_start(*action_bar_label);
      action_bar->pack_end(*action_bar_button);

      box->append(*box_label);
      box->append(*action_bar);

      box_label->set_vexpand(true);
      box_label->set_hexpand(true);

      add_window(*window);
    }
};

int main(int argc, char** argv) {
  return App::create()->run(argc, argv);
}